src/main/java/org/jabref/cli/ArgumentProcessor.java

```diff
@@ -239,6 +253,114 @@ public boolean hasParserResults() {
+
+        ParserResult pr = loaded.get(loaded.size() - 1);
+        BibDatabaseContext databaseContext = pr.getDatabaseContext();
+        BibDatabase dataBase = pr.getDatabase();
+
+        XmpPdfExporter xmpPdfExporter = new XmpPdfExporter(preferences.getXmpPreferences());
+
+        if (filesAndCitekeys.equals("all")) {
+            for (BibEntry entry : dataBase.getEntries()) {
+                writeXMPtoPDFsOfEntry(databaseContext, dataBase, entry, preferences, xmpPdfExporter);
+            }
+            return;
+        }
+
+        Vector<String> citeKeys = new Vector<>();
+        Vector<String> pdfs = new Vector<>();
+        for (String fileOrCiteKey : filesAndCitekeys.split(",")) {
+            if (fileOrCiteKey.toLowerCase(Locale.ROOT).endsWith(".pdf")) {
+                pdfs.add(fileOrCiteKey);
+            } else {
+                citeKeys.add(fileOrCiteKey);
+            }
+        }
+
+        writeXMPtoPdfByCitekey(databaseContext, dataBase, citeKeys, preferences, xmpPdfExporter);
+        writeXMPtoPdfByFileNames(databaseContext, dataBase, pdfs, preferences, xmpPdfExporter);
+
+    }
+
+    private void writeXMPtoPDFsOfEntry(BibDatabaseContext databaseContext, BibDatabase dataBase, BibEntry entry, JabRefPreferences preferences, XmpPdfExporter xmpPdfExporter) {
+        boolean writtenToAFile = false;
+
+        for (LinkedFile file : entry.getFiles()) {
+            if (file.getFileType().toLowerCase(Locale.ROOT).equals("pdf")) {
+                Optional<Path> filePath = file.findIn(databaseContext, preferences.getFilePreferences());
+                if (filePath.isEmpty()) {
+                    System.err.println(Localization.lang("Skipped - PDF %0 does not exist", filePath.get()));
+                } else {
+                    try {
+                        xmpPdfExporter.export(databaseContext, filePath.get(), preferences.getDefaultEncoding(), Arrays.asList(entry));
+                        writtenToAFile = true;
+                    } catch (Exception e) {
+                        System.err.println(Localization.lang("Error writing XMP content: %0", filePath.get()));
+                    }
+                }
+            }
+        }
+
+        if (!writtenToAFile) {
+            System.err.println(Localization.lang("Skipped - No PDF linked to entry %0", entry.getCitationKey()));
+        }
+    }
+
+    private void writeXMPtoPdfByCitekey(BibDatabaseContext databaseContext, BibDatabase dataBase, Vector<String> citeKeys, JabRefPreferences preferences, XmpPdfExporter xmpPdfExporter) {
+        for (String citeKey : citeKeys) {
+            Optional<BibEntry> entry = dataBase.getEntryByCitationKey(citeKey);
+            if (entry.isEmpty()) {
+                System.err.println(Localization.lang("Cannot find %0 in library.", citeKey));
+                return;
+            }
+            writeXMPtoPDFsOfEntry(databaseContext, dataBase, entry.get(), preferences, xmpPdfExporter);
+        }
```

Comment:
We love streams. We love foreach. 😅

Reply:
Good Idea

Reply:
All but not foreach, because it's hard to debug

Reply:
Hm. I played around with all the loops I introduced and can't figure out a way to replace them with streams while avoiding forEach without suppressing the warning messages I am printing currently.
For the example snippet you highlighted, I could change it to something like
`citeKeys.stream().map(citeKey -> dataBase.getEntryByCitationKey(citeKey)).filter(Optional<>::isPresent).map(Optional<>::get).forEach(bibEntry -> writeXMPtoPDFsOfEntry(...))`
But:
- that still uses forEach (I guess such a use would be ok, how else would one do this?)
- I cannot print an exception to alert the user that an invalid cite key was given (not found in library).
- 
Also, I need to catch some exceptions (i.e. the one thrown by the XmpPdfExporter's export method) so for that I would also need forEach (AFAIK).

Can you give me some pointers how to clean up these loops?

Reply:
I would not really use streams if it doesn't make sense or you have exception handling

Reply:
Seems like @Siedlerchr does not like streams. 😉

Reply:
Streams are fine, but if you need exception handling its not always ideal